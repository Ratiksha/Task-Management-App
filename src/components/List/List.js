import React, { Component } from 'react';
import './List.scss';
import Card from '../Card/Card';
import Modal from '../Modal/Modal'
import AddTaskForm from '../AddTaskForm/AddTaskForm'
import DeleteForm from '../DeleteForm/DeleteForm'
import AddListForm from '../AddListForm/AddListForm'
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt, faPencilAlt } from "@fortawesome/free-solid-svg-icons";


export default class List extends Component{
    constructor(props){
        super(props);
        this.state = {
            showModal: false,
            showEditForm: false,
            showDeleteForm: false
        }
        this.handleKeyPress = this.handleKeyPress.bind(this);
    }

    componentDidMount(){
        document.addEventListener('keydown', this.handleKeyPress);
    }

    handleKeyPress(event) {
        if (event.keyCode === 13) {
            document.activeElement.click();
        }
      }

    showModal = (formType) => {
        this.toggleModal();
        if(formType === 'edit'){
            this.setState({ 
                showEditForm: true,
                showDeleteForm: false
            });
        }
        else if(formType === 'delete') {
            this.setState({ 
                showDeleteForm: true,
                showEditForm: false
            });
        }
    }
    
    hideModal = () => {
        this.toggleModal();
    }

    toggleModal = () => {
        this.setState({ 
            showModal: !this.state.showModal,
            showDeleteForm: false,
            showEditForm: false
         });
    }

    render() {
        const { lists, list, addTask, editTask, DeleteTask, editList, DeleteList, onTaskDragStart, onTaskDragOver, onTaskDrop, onListDragStart  } = this.props;
        const cards = list && list.cards.map((card, index) => (
            <li key={index}>
                <Card lists={lists} card={card} editTask={editTask} DeleteTask={DeleteTask} listId={list.listId} onTaskDragStart={onTaskDragStart}/>
            </li>
        ));

        return (
            <div draggable="true" id={[ list && list.listId]} onDragStart={onListDragStart}>
                <h3 className="header">{list && list.name}
                <div className="button-group">
                    <button id="list-edit" onClick={() => this.showModal('edit')}><FontAwesomeIcon icon={faPencilAlt}/></button>
                    <button id="list-delete" onClick={() => this.showModal('delete')}><FontAwesomeIcon icon={faTrashAlt}/></button>
                </div>
                </h3>
                <ul className="list" onDragOver={onTaskDragOver} onDrop={onTaskDrop}>
                    {cards}
                    <li>
                        <AddTaskForm lists={lists} listId={list && list.listId} addTask={addTask} hideModal={this.hideModal}/>
                    </li>
                </ul>
                <Modal show={this.state.showModal} handleClose={this.hideModal}>
                    {this.state.showEditForm && <AddListForm lists={lists} listId={list && list.listId} editList={editList} name={list && list.name} hideModal={this.hideModal}/>}
                   {this.state.showDeleteForm && <DeleteForm listId={list && list.listId} DeleteList={DeleteList} name={list && list.name} hideModal={this.hideModal}/>}
                </Modal>
            </div>
        )
    }
}