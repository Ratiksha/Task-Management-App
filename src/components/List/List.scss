@import '../variables.scss';
.header{
    padding: 8px 10px;
    margin: 0;
    position: relative;
    background: var(--color);
    color: $white;
    border-top-right-radius: 5px;
    border-top-left-radius: 5px;
    .button-group{
        position: absolute;
        top: 7px;
        right: 10px;
        button{
            background-color: transparent;
            padding: 0;
        }
        svg{
            color: $white;
            margin: 5px;
            font-size: 10px;
            cursor: pointer;
        }
    }
}

.list{
    list-style-type: none;
    padding: 20px 8px 20px;
    border-radius: 2px;
    min-height: 218px;
}