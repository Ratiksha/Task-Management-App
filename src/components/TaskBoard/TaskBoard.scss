@import '../variables.scss';
:root{
    --color: #227fba;
}

html{
    background: #ebf0f0;
    button, .button{
        border: none;
        color: $white;
        padding: 6px 12px;
        text-align: center;
        text-decoration: none;
        display: inline-block;
        cursor: pointer;
        &.green{
        background: $green;
        }
        &.grey{
             background: #9E9E9E;;
        }
    }
}
.taskBoard{
    .main-header{
        h1{
            color: $white;
            border-bottom: 1px solid #e4e4e4;
            padding: 20px;
            margin: 0;
            background: var(--color);
            line-height: 1.5;
        }
        svg{
            font-size: 35px;
            color: var(--color);
            cursor: pointer;
        }
    }
        .sub-header{
            padding: 20px;
            ul{
                display: inline;
                float: right;
                margin: 0;
                .colors{
                    width: 25px;
                    height: 25px;
                    display: inline-block;
                    border-radius: 50%;
                    cursor: pointer;
                    margin: auto 5px;
                    box-shadow: 1px 1px 7px 2px #0000001c;
                }
                #blue{
                    background-color: $blue;
                }
                #yellow{
                    background-color: $yellow;
                }
                #purple{
                    background-color: $purple;
                }
            }
        }
       
    
    .lists{
        display:flex;
        flex-direction: row;
        flex-wrap: wrap;
        align-items: flex-start;
        justify-content: flex-start;
        margin: 0 auto;
        padding: 20px;
        list-style-type: none;
        @media (min-width: 768px) and (max-width: 1999px) { 
            width: 768px;
        }
        @media only screen and (min-width: 1200px) {
            width: 1200px;
        }
        .wrapper{
            background: $white;
            margin: 30px;
            border-radius: 5px;
            height: 300px;
            width: 90%;
            box-shadow: 0px 3px 14px 2px #0000001c;
            overflow-y: auto;
            .modal{
                z-index: 9;
            }
            @media (min-width: 768px) and (max-width: 1999px) { 
                width: 40%;
            }
            @media only screen and (min-width: 1200px) {
                width: 28%;
            }
        }
    }
    input[type=text] {
        padding: 5px 20px;
        margin: 8px 0;
        margin-right: 15px;
      }
}

