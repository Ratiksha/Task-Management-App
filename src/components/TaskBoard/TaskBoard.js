import React, { Component } from 'react';
import './TaskBoard.scss';
import List from '../List/List';
import Modal from '../Modal/Modal'
import AddListForm from '../AddListForm/AddListForm'
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlusSquare } from "@fortawesome/free-solid-svg-icons";

export default class TaskBoard extends Component{
    constructor(props) {
        super(props);
        this.state = {
            lists:[
                {
                    name: 'TO DO',
                    listId: 0,
                    cards: [{
                        taskName: 'Pay Electricity Bill',
                        list: 0,
                        taskId: 0
                    },
                    {
                        taskName: 'Make Grocery List',
                        list: 0,
                        taskId: 1
                    }]
                },
                {
                    name: 'In Progress',
                    listId: 1,
                    cards: [{
                        taskName: 'Iron Cloths',
                        list: 1,
                        taskId: 0
                    },]
                },
                {
                    name: 'Done',
                    listId: 2,
                    cards: [{
                        taskName: 'Buy Running Shoes',
                        list: 2,
                        taskId: 0
                    }]
                }
            ],
            show: false,
            taskdragInfo: "",
            listdragInfo: ""
        }

        this.handleKeyPress = this.handleKeyPress.bind(this);
        
    }

    componentDidMount(){
        const localStorageData = localStorage.getItem('lists');
        if(localStorageData){
            this.setState({
                lists: JSON.parse(localStorageData)
            })
        }
        const themeColor = localStorage.getItem('themeColor');
        document.documentElement.style.setProperty('--color', JSON.parse(themeColor));
        document.addEventListener('keydown', this.handleKeyPress);
    }

    handleKeyPress(event) {
        if (event.keyCode === 13) {
            document.activeElement.click();
        }
      }

    changeTheme = (color) =>{
        document.documentElement.style.setProperty('--color', color);
        localStorage.setItem('themeColor', JSON.stringify(color))
    }
    

  onTaskDragStart = (event, listID) => {
    event.stopPropagation();
    const taskdragInfo = {
      taskId: event.currentTarget.id,
      listID: listID
    }
    this.setState({
        taskdragInfo
    })
  }

  onTaskDragOver = (event) => {
    event.preventDefault();
  }

  onTaskDrop = (event, listID) => {
    event.stopPropagation();
    const { lists, taskdragInfo } = this.state
    var cardsArray = []
    for (var i = 0; i < lists.length; i++) {
        if(parseInt(taskdragInfo.listID) === lists[i].listId){
            cardsArray = lists[i].cards
        }
    }
    const taskCard = cardsArray.find(card => card.taskId === parseInt(taskdragInfo.taskId))
    const indexOfCard = cardsArray.findIndex(card => card.taskId === parseInt(taskdragInfo.taskId))
    for (var k = 0; k < lists.length; k++) {
        if(parseInt(taskdragInfo.listID) === lists[k].listId){
            lists[k].cards.splice(indexOfCard, 1)
        }
    }
    for (var j = 0; j < lists.length; j++) {
        if(parseInt(listID) === lists[j].listId){
            lists[j].cards.push({...taskCard, list: parseInt(listID), taskId: new Date().valueOf() });
        }
    }
    this.setState({
      lists
    });
    localStorage.setItem('lists', JSON.stringify(lists));
    
  }

  onListDragStart = (event, listID) => {
    const listdragInfo = {
      listID: listID
    }
    this.setState({
        listdragInfo
    })
  }

  onListDragOver = (event, listId) => {
    event.preventDefault();
  }

  onListDrop = (event, listID) =>{
    const { lists, listdragInfo } = this.state
    var updatedparsedData =  lists.filter(list => list.listId !== parseInt(listdragInfo.listID))
    const taskList = lists.find(list => list.listId === parseInt(listdragInfo.listID))
    const indexOfList = lists.findIndex(list => list.listId === parseInt(listID))
    updatedparsedData.splice(indexOfList, 0, taskList)
    this.setState({
        lists: updatedparsedData
      });
      localStorage.setItem('lists', JSON.stringify(updatedparsedData));

  }

    addList(name){
        const { lists } = this.state
        const newList = {
            name: name,
            listId: new Date().valueOf(),
            cards: []
        }
        lists.push(newList)
        this.setState({
            lists
        })
        localStorage.setItem('lists', JSON.stringify(lists))
    }

    editList(name, listId){
        const { lists } = this.state
        for (var i = 0; i < lists.length; i++) {
            if(listId === lists[i].listId){
                lists[i].name = name ;
            }
          }
          this.setState({
            lists
        })
        localStorage.setItem('lists', JSON.stringify(lists))
    }

    DeleteList(listId){
        const { lists } = this.state
        const updatedList = lists.filter(list => list.listId !== listId)
          this.setState({
            lists: updatedList
        })
        localStorage.setItem('lists', JSON.stringify(updatedList))
    }

    addTask(taskName, listId){
        const { lists } = this.state
        const newTask = {
            taskName,
            list: listId,
            taskId: new Date().valueOf()
        }
        for (var i = 0; i < lists.length; i++) {
            if(listId === lists[i].listId){
                lists[i].cards.push(newTask);
            }
        }
        this.setState({
            lists
        })
        localStorage.setItem('lists', JSON.stringify(lists))
    }

    editTask(taskName, listId, taskId){
        const { lists } = this.state
        for (var i = 0; i < lists.length; i++) {
            if(listId === lists[i].listId){
                for (var j = 0; j < lists[i].cards.length; j++) {
                    if(lists[i].cards[j].taskId === taskId){
                        lists[i].cards[j].taskName = taskName 
                    }
                }
            }
          }
          this.setState({
            lists
        })
        localStorage.setItem('lists', JSON.stringify(lists))
    }

    DeleteTask(listId, taskId){
        const { lists } = this.state
        for (var i = 0; i < lists.length; i++) {
            if(listId === lists[i].listId){
                for (var j = 0; j < lists[i].cards.length; j++) {
                    if(lists[i].cards[j].taskId === taskId){
                        lists[i].cards =  lists[i].cards.filter(card => card.taskId !== taskId)
                    }
                }
            }
          }
          this.setState({
            lists
        })
        localStorage.setItem('lists', JSON.stringify(lists))
    }

    showModal = () => {
        this.setState({ show: true });
      };
    
      hideModal = () => {
        this.setState({ show: false });
      };

    render() {
        const lists = this.state.lists.map((list, index) => (
            <li key={index} className="wrapper" onDragOver={(event, listId)=>this.onListDragOver(event, list.listId)} onDrop={(event)=>this.onListDrop(event, list.listId)} tabIndex="0">
                <List list={list}
                lists={this.state.lists}
                addList={(name) => this.addList(name)}
                editList={(name, listId) => this.editList(name, listId)}
                DeleteList={(listId) => this.DeleteList(listId)}

                addTask={(taskName, listId) => this.addTask(taskName, listId)}
                editTask={(taskName, listId, taskId) => this.editTask(taskName, listId, taskId)}
                DeleteTask={(listId, taskId) => this.DeleteTask(listId, taskId)}

                onTaskDragStart={(event) => this.onTaskDragStart(event, list.listId)}
                onTaskDragOver={(event) => this.onTaskDragOver(event)} 
                onTaskDrop={(event) => {this.onTaskDrop(event, list.listId)}}
                
                onListDragStart={(event) => {this.onListDragStart(event, list.listId)}}
                />
            </li>
        ));

        return (
            <div className="taskBoard">
                <div className="main-header">
                    <h1>Task Management Board</h1>
                    <div className="sub-header">
                        <ul>
                            <li className="colors" id="blue" onClick={() => this.changeTheme('#227fba')} tabIndex="0"></li>
                            <li className="colors" id="yellow" onClick={()=>this.changeTheme('#f59e01')} tabIndex="0"></li>
                            <li className="colors" id="purple" onClick={()=>this.changeTheme('#540073')} tabIndex="0"></li>
                        </ul>
                        <button id="add-list" onClick={this.showModal}><FontAwesomeIcon icon={faPlusSquare}/></button>
                    </div>
                    </div>
                <ul className="lists" id="list">{lists}</ul>
                <Modal show={this.state.show} handleClose={this.hideModal} id="listModal">
                    <AddListForm lists={this.state.lists} addList={(name) => this.addList(name)} hideModal={this.hideModal}/>
                </Modal>
            </div>
        )
    }
}