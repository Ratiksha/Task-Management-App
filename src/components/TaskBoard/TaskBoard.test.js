import React from 'react';
import { shallow, mount } from 'enzyme';
import TaskBoard from './TaskBoard';
import List from '../List/List'

jest.mock('../List/List', () => 'List');

let wrapped = shallow(<TaskBoard />);
describe('TaskBoard', () => {
  it('should render the TaskBoard Component correctly', () => {   
    expect(wrapped).toMatchSnapshot();
  });

  it('When blue theme button is clicked', () => {
    const color = '#227fba'
    const blueButton = wrapped.find('#blue');
    blueButton.simulate('click');
    expect(color).toBe('#227fba');
  });

  it('When yellow theme button is clicked', () => {
    const color = '#f59e01'
    const yellowButton = wrapped.find('#yellow');
    yellowButton.simulate('click');
    expect(color).toBe('#f59e01');
  });

  it('When purple theme button is clicked', () => {
    const color = '#540073'
    const purpleButton = wrapped.find('#purple');
    purpleButton.simulate('click');
    expect(color).toBe('#540073');
  });

  it('When add list button is clicked', () => {
    const showModal = jest.fn();
    showModal();
    wrapped.find('#add-list').simulate('click')
    expect(showModal).toHaveBeenCalled();
  });

});