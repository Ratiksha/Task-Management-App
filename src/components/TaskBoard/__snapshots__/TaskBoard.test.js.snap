// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskBoard should render the TaskBoard Component correctly 1`] = `ShallowWrapper {}`;
